Add changes in HVGs and clustering to change report (#871)
AlexsLemonade/scpca-nf: scripts/compare-metrics/compare-metrics-template.rmd

    ...
  )
}

## Functions for cluster size comparison
make_clusters <- function(x) {
  # convert a vector of cluster sizes to a vector of cluster assignments
  # x is a vector of cluster sizes
  # returns a vector of cluster assignments
  rep(seq_along(x), times = x)
}

pair_sizes <- function(a_sizes, b_sizes) {
  # find the difference between every pair of a and b,
  # then choose from smallest to largest, eliminating used pairs as you go

  match_df <- expand.grid(
    a_i = seq_along(a_sizes),
    b_i = seq_along(b_sizes)
  ) |>
    dplyr::mutate(
      a = a_sizes[a_i],
      b = b_sizes[b_i],

>       share_frac = pmin(a, b) / pmax(a, b),

Reviewer: ```suggestion
```
I don't think you use this at all in this function?

Author: I do; that is what I am sorting by to decide which samples to pick first.

Reviewer: Wow I'm blind and clearly didn't see the next line.

      used = FALSE
    ) |>
    dplyr::arrange(desc(share_frac))

  # pre-allocate result vectors
  new_a <- integer(min(length(a_sizes), length(b_sizes)))
  new_b <- integer(min(length(a_sizes), length(b_sizes)))
  pair_count <- 0
  # loop until all pairs are used

>   while (!all(match_df$used)) {

Reviewer: Generally, this took me a while to figure out exactly what's going on here, so can you add some more descriptive comments?

Author: Yes, I will add some more comments.

    # Find the first unused pair
    idx <- which(!match_df$used)[1]
    # add indexes to result vectors
    pair_count <- pair_count + 1
    new_a[pair_count] <- match_df$a_i[idx]
    new_b[pair_count] <- match_df$b_i[idx]
    # remove the first row and any rows with those indexes
    # Mark used pairs
    used_a <- match_df$a_i == new_a[pair_count]
    used_b <- match_df$b_i == new_b[pair_count]
    match_df$used[used_a | used_b] <- TRUE
  }
  # add any missing indexes to the end
  new_a <- c(new_a, setdiff(seq_along(a_sizes), new_a))
  new_b <- c(new_b, setdiff(seq_along(b_sizes), new_b))
  # return values
  return(list(a_sizes[new_a], b_sizes[new_b]))
}

> max_ari <- function(a_sizes, b_sizes) {

Reviewer: This approach to calculating the max ARI seems reasonable to me. I don't really have a better idea and I think we will know more once we put this into practice if it's a good metric.

    library_id,
    added_removed
  )

> # filter metrics down libraries in both sets for the remainder of the notebook

Reviewer: ```suggestion
# filter metrics down to libraries in both sets for the remainder of the notebook
```

      hv_genes.ref, hv_genes.comp,
      \(x, y){
        length(intersect(x, y)) / length(union(x, y))
      }
    ),
    hvg_rank_cor = purrr::map2_dbl(
      hv_genes.ref, hv_genes.comp,
      \(x, y) {
        cor(
          seq_along(x),
          # get the ordering of a elements in b
          as.integer(factor(y, levels = x)),
          use = "complete.obs", # discard
        )
      }
    )
  ) |>
  dplyr::select(
    project_id,
    sample_id,
    library_id,
    starts_with(c("hv_genes", "hvg"))
  )

hvg_changes <- hvg_stats |>
  dplyr::filter(!hvg_identical)

has_hvg_changes <- nrow(hvg_changes) > 0
```

```{r}
# HVG change summary
hvg_summary <- hvg_stats |>
  dplyr::summarise(
    n = dplyr::n(),
    n_changed = n - sum(hvg_identical),
    mean_change = mean(hvg_n_changed[which(!hvg_identical)]),
    mean_change_frac = mean(hvg_change_frac[which(!hvg_identical)]),
    mean_jaccard = mean(hvg_jaccard[which(!hvg_identical)]),
    mean_cor = mean(hvg_rank_cor[which(!hvg_identical)])
  )

knitr::asis_output(glue::glue_data(
  hvg_summary,
  "Of the **{n}** libraries, **{n_changed}** {ifelse(n_changed == 1, 'has', 'have')} changes in highly variable genes.

>   Among the libraries with changes, the mean number of genes changed is **{mean_change}**, which is a mean percentage change of **{round(mean_change_frac * 100, digits = 2)}%**.

Reviewer: I think we might want to update this text (and maybe the table to?) to be explicit about which libraries have changes in the genes themselves vs. which libraries only have changes in the order.

Author: This is kind of present in the table; if the Jaccard index is 1, then the gene set is unchanged and only the order has changed. I kind of doubt this will ever happen in practice; I expect the bottom end of the list to change with some frequency. One thing I had thought about (and never implemented, as I'm only remembering the passing thought now) was calculating these stats for subsets of the HVGs; I can imagine that the top 500 or 1000 genes might be more stable, which might be nice to see at a glance.

Reviewer: > This is kind of present in the table; if the Jaccard index is 1, then the gene set is unchanged and only the order has changed. I kind of doubt this will ever happen in practice; I expect the bottom end of the list to change with some frequency. One thing I had thought about (and never implemented, as I'm only remembering the passing thought now) was calculating these stats for subsets of the HVGs; I can imagine that the top 500 or 1000 genes might be more stable, which might be nice to see at a glance.

Maybe we should add a note that the table shows only those that have changes, either in the gene content or the order of the genes as a reminder. Unless we have that stated somewhere already?

    hvg_jaccard,
    hvg_rank_cor
  ) |>
  tidyr::pivot_longer(
    cols = everything(),
    names_to = "metric",
    values_to = "value"
  )

ggplot(hvg_plot_df, aes(x = value)) +
  geom_histogram(bins = 20) +
  facet_wrap(vars(metric), ncol = 3, scales = "free_x") +
  labs(
    title = "Histograms of change metrics for highly variable genes",
    x = "",
    y = "Number of libraries"
  ) +
  theme_bw()
```

```{r, eval=has_hvg_changes}
knitr::asis_output("#### Detail")
```

```{r, eval=has_hvg_changes}
# detail table of hvg changes

hvg_changes <- hvg_changes |>
  dplyr::select(
    project_id,
    sample_id,
    library_id,
    hvg_n_changed,
    hvg_change_frac,
    hvg_jaccard,
    hvg_rank_cor
  )

report_table(
  hvg_changes,
  colnames = c(
    "Project",
    "Sample",
    "Library",
    "Number of genes changed",
    "% change",
    "Jaccard index",
    "Rank correlation"
  )
) |>
  DT::formatPercentage(c("hvg_change_frac"), 2)
```

### Clustering {.tabset}

Report changes in the number of clusters, as well as changes in the sizes of the assigned clusters.
We will ignore changes in the ordering of clusters by sorting by cluster size before comparisons.

> As a summary statistic, we might use adjusted Rand index (ARI) or adjusted mutual information (AMI) to compare the two sets of clusters, reporting when we are well below 1.

Author: I sure meant to!

```{r}
# compare clusters
cluster_stats <- metrics_df |>
  dplyr::mutate(
    cluster_setequal = purrr::map2_lgl(cluster_sizes.ref, cluster_sizes.comp, setequal),
    cluster_n_change = purrr::map2_int(
      cluster_sizes.ref,
      cluster_sizes.comp,
      \(x, y) {
        length(y) - length(x)
      }
    ),

>     cluster_n_change_frac = cluster_n_change / purrr::map_int(cluster_sizes.ref, length),

Reviewer: Is this the total number of clusters that have changed divided by the total number of clusters?

Author: This is the change in the number of clusters, relative to the starting number of clusters. I definitely need to clarify!

    paired_clusters = purrr::map2(cluster_sizes.ref, cluster_sizes.comp, pair_sizes),
    cluster_sizes.ref = purrr::map(paired_clusters, \(x) x[[1]]),
    cluster_sizes.comp = purrr::map(paired_clusters, \(x) x[[2]]),
    cluster_max_ari = purrr::map2_dbl(cluster_sizes.ref, cluster_sizes.comp, max_ari)
  ) |>
  dplyr::select(
    project_id,
    sample_id,
    library_id,
    starts_with("cluster")
  )

cluster_changes <- cluster_stats |>
  dplyr::filter(!cluster_setequal)

has_cluster_changes <- nrow(cluster_changes) > 0
```

```{r}
# cluster change overall summary
cluster_summary <- cluster_stats |>
  dplyr::summarise(
    n = dplyr::n(),
    n_changed = n - sum(cluster_setequal),
    mean_n_change = mean(abs(cluster_n_change[which(!cluster_setequal)])),

>     mean_n_change_frac = mean(abs(cluster_n_change_frac[which(!cluster_setequal)])),

Reviewer: I think I need some clarity on what % change and mean % change is here? Is it the number of clusters that have different sizes? Or is it how different are the clusters with different sizes?

    mean_max_ari = mean(cluster_max_ari[which(!cluster_setequal & cluster_n_change == 0)])
  )

if (!has_cluster_changes) {
  knitr::asis_output(
    "There are no changes in clustering between the reference and comparison runs."
  )
} else {
  knitr::asis_output(glue::glue_data(
    cluster_summary,
    "Of the **{n}** libraries, **{n_changed}** {ifelse(n_changed ==1, 'has', 'have')} changes in clustering.<br>

>     Among the libraries with changes, the mean change in the number of clusters is **{mean_n_change}**, which corresons  a mean percentage change of **{round(mean_n_change_frac * 100, digits = 2)}%**.

Reviewer: ```suggestion
    Among the libraries with changes, the mean change in the number of clusters is **{mean_n_change}**, which corresonds to a mean percentage change of **{round(mean_n_change_frac * 100, digits = 2)}%**.
```

    The mean `maxARI` for the changed clusterings is **{round(mean_max_ari, digits = 2)}**.
    This is an approximate maximum value calculated based on assuming the best correspondence in sizes among the clusters.
    Note that this only includes comparisons where the number of cells is unchanged between the two runs.
    "
  ))
}
```

```{r, eval=has_cluster_changes}
knitr::asis_output("#### Summary")
```

```{r, eval=has_cluster_changes}
# cluster change plots
cluster_plot_df <- cluster_changes |>
  dplyr::select(cluster_n_change_frac, cluster_max_ari) |>
  tidyr::pivot_longer(
    cols = everything(),
    names_to = "metric",
    values_to = "value"
  )

cluster_plot <- ggplot(cluster_plot_df, aes(x = value)) +
  geom_histogram(bins = 20) +
  facet_wrap(vars(metric), scales = "free_x") +
  labs(
    title = "Histograms of change metrics for clustering",
    x = "",
    y = "Number of libraries"
  ) +
  theme_bw()

suppressWarnings(print(cluster_plot))
```

```{r}
# plot top changed clusters
plot_df <- cluster_changes |>
  dplyr::arrange(desc(cluster_max_ari)) |>

>   head(6) |> # only plot the top 6

Reviewer: I wonder here if it actually makes sense to show all the libraries that have changes in the histogram but only the top ones in the stacked bar chart? I feel like this could change slightly across the board for a lot of libraries, so it would be helpful to see if there are libraries that are outliers or not.

Author: Can you clarify what you mean here? The current histograms are all changes, and the stacked plots are only the ones with the highest max ARI. There definitely needs to be some text explaining that, but I was wondering if you meant that we should _not_ do that?

Reviewer: I think I was confused because the histograms only show 3 libraries for the `cluster_max_ari`. Is it just that the number of libraries with changes is only 6 here anyways? Or did you accidentally filter them out somewhere?

Reviewer: I see that there are only 3 for the ARI because some of them have NA, do we expect that?

Author: Yeah, there are only 6 libraries with changes. And yes, we expect ARI to be `NA` for some libraries, because we can't calculate ARI if the number of cells has changed. Well, I'm sure I could do something but I didn't have any good ideas.

Reviewer: > And yes, we expect ARI to be NA for some libraries, because we can't calculate ARI if the number of cells has changed. 

Also worth a note in the text!

Reviewer: Alternatively, if the number of cells has changed then the clustering should have definitely changed so we could remove those as we do with the reads table?

Author: I think we want to keep something there, because if we have a small change in the number of cells we would still want to know if the clustering was affected. This makes me think I should still try to do something if the cell count isn't equal. Maybe scaling down the counts somehow...

  dplyr::select(project_id, sample_id, library_id, cluster_sizes.ref, cluster_sizes.comp) |>
  tidyr::pivot_longer(
    cols = starts_with("cluster_sizes"),
    names_to = "set",
    names_pattern = ".+\\.(ref|comp)$",
    values_to = "cluster_sizes"
  ) |>
  tidyr::unnest_longer(
    cluster_sizes,
    indices_to = "cluster",
    values_to = "size"
  ) |>
  dplyr::mutate(cluster = as.factor(cluster))

# rotating color palette
cluster_colors <- palette.colors(
  palette = "Tableau",
  n = length(levels(plot_df$cluster)),
  recycle = TRUE
)

ggplot(plot_df, aes(x = set, y = size, fill = cluster)) +
  geom_col(position = position_stack(reverse = TRUE)) +
  facet_wrap(vars(library_id), scales = "free_y") +
  labs(
    title = "Cluster sizes for the libraries with the largest max ARI changes",
    x = "",
    y = "Cell count",
  ) +
  scale_fill_discrete(type = cluster_colors) +
  theme_bw() +
  theme(legend.position = "none")
```

```{r, eval=has_cluster_changes}
knitr::asis_output("#### Detail")
```

```{r, eval=has_cluster_changes}

> # Detail table of cluster changes

Reviewer: It's confusing to me how libraries that have 0 clusters changed are showing up in this table? I would expect only those with a different number of clusters changed?

Author: This is a labeling problem. It should say "Change in number of clusters". Or "change in cluster number" or something like that.

My initial thought is to use the Euclidean distance between the sets of cell type assignments here (since the assignment values do matter), though we could also look at ARI or AMI.

```{r}
# compare cell type assignments

celltypes_changes <- function(ref_celltypes, comp_celltypes) {
  type_matrix <- dplyr::bind_rows(a, b) |> as.matrix()
  rownames(type_matrix) <- c("ref", "comp")
  type_matrix[is.na(type_matrix)] <- 0

  added <- colnames(type_matrix)[which(type_matrix["ref"] == 0 && type_matrix["comp"] > 0)]
  removed <- colnames(type_matrix)[which(type_matrix["comp"] == 0 && type_matrix["ref"] > 0)]
  n_changes <- dist(type_matrix, method = "manhattan")[1]

>   euclidean_distance <- dist(type_matrix, method = "euclidean")[1]

Reviewer: I don't know if you meant to add this in here? I think I would remove it.

Author: I kind of did... I was originally doing all three of these parts together, but then decided it was better to finish the first two. But I'll remove it before merging you you can look at it fresh in the next PR
